### DISCUSSION TITLE
Cloud-Native Orc8r Secrets

### DISCUSSION THREAD

**Discussion by hcgatewood** [magma/magma#5728] [Ideas]
# Cloud-Native Orc8r Secrets

This document describes a plan for moving Orc8r to cloud-native secrets management, resolving some open issues with Orc8r secrets in the process.

Review requested from
- Cloud: @emakeev, @karthiksubraveti, @mpgermano 
- Gateway: @amarpad, @ulaskozat 
- Cloud-native secrets: @arunuke, @ShubhamTatvamasi (PR owner)

## Summary

Thanks to https://github.com/magma/magma/pull/5584, we have the opportunity to move to a cloud-native approach for Orc8r secrets management. This document expands on the work in that PR by aligning on a story for how Orc8r's cloud-native secrets management will affect other areas of the project.

### Wins

- Automatically generate secrets on first deployment
- Automatically rotate certificates
- Automatically rotate secrets
- Handle Magma's "10-year problem"

### Open questions

- Are we okay with removing support for self-signed certs in favor of defaulting to automatic Let's Encrypt certs
- What are reasonable renewal periods for the certs and keys listed in [Solution](#solution) section.

## Context

Read [Orc8r Security Overview](https://magma.github.io/magma/docs/next/orc8r/architecture_security) and [Orc8r Security In-Depth](https://magma.github.io/magma/docs/next/orc8r/dev_security) for an end-to-end understanding of Orc8r's secrets.

### Orc8r overview

![orc8r_proxying](https://user-images.githubusercontent.com/8029544/112415688-40ad3700-8cf2-11eb-9330-485368150dae.png)

### Trust anchors

Orc8r utilizes the following trust anchors across the Orc8r-gateway interface

- `rootCA.pem` is the root CA for Orc8r's TLS certificate
- `fluentd.pem` is the root CA for Orc8r's *fluentd* TLS certificate
- `certifier.pem` is the root CA for gateways' TLS certificates

### 10-year problem

Currently, Orc8r supports all 3 anchors being self-signed. This is fine for `certifier.pem`, but it requires manually copying both `rootCA.pem` and `fluentd.pem` onto *each gateway connecting to that Orc8r*.

This creates Magma's "10-year problem" -- when self-signed, these certs are generated with a [validity period of 10 years](https://github.com/magma/magma/blob/50816a97df8f3873bdb5059d556230ab0d2831ba/orc8r/cloud/deploy/scripts/self_sign_certs.sh#L41), and the current architecture provides no affordances for updating these certs at the gateways, aside from manually copying them over. (Side note: this problem still exists for non-self-signed certs, with issuer-specific timeline.)

Aside from being a sticky hassle every 10 years, this pitfall entails additional issues for all deployments

- Requires manual intervention to update `controller.crt` (Orc8r's TLS cert) every 3 years
- Precludes good security hygiene of [max validity period of 1 year](https://thehackernews.com/2020/09/ssl-tls-certificate-validity-398.html).
- Precludes good security hygiene of [max secret lifetime of 1 year](https://www.venafi.com/blog/how-frequently-should-you-rotate-pki-certificates-and-keys)

### Upshot

Issues
- 🚨 [A] Can't automatically renew and distribute `rootCA.pem` ([every 10 years for self-signed](https://github.com/magma/magma/blob/509e5b4bcdb68b0021044fb1dc8cf47554446b8f/orc8r/cloud/go/tools/accessc/handlers/common_add.go#L52-L53))
- 🚨 [B] Can't automatically renew and distribute `fluentd.pem` ([every 10 years for self-signed](https://github.com/magma/magma/blob/509e5b4bcdb68b0021044fb1dc8cf47554446b8f/orc8r/cloud/deploy/scripts/create_application_certs.sh#L54))
- 🚨 [C] Can't automatically renew `controller.crt` ([every 3 years for self-signed](https://github.com/magma/magma/blob/50816a97df8f3873bdb5059d556230ab0d2831ba/orc8r/cloud/deploy/scripts/self_sign_certs.sh#L59))

Nice-to-have
- ⚠️ [D] Can't automatically rotate ephemeral keys ([`bootstrapper.key`](https://github.com/magma/magma/blob/509e5b4bcdb68b0021044fb1dc8cf47554446b8f/orc8r/cloud/deploy/scripts/create_application_certs.sh#L47))
- ⚠️ [E] Can't automatically rotate non-ephemeral keys
- ⚠️ [F] [Secrets are manually generated during install](https://magma.github.io/magma/docs/next/orc8r/deploy_install#assemble-certificates)
- ⚠️ [G] [Secrets management is coupled with AWS](https://github.com/magma/magma/blob/e8a4f5725c69361b205459171016dc82dc1cfe89/orc8r/cloud/deploy/terraform/orc8r-helm-aws/scripts/create_orc8r_secrets.py#L44-L46)

## Solution

[cert-manager](https://cert-manager.io/docs/) is a [CNCF sandbox project](https://www.jetstack.io/blog/cert-manager-cncf/) that provides a cloud-native, all-K8s mechanism for generating and rotating secrets. Aside from streamlining the Orc8r install process, it opens the way for resolving the other listed issues.

### Cert-manager overview

![cert-manager overview](https://user-images.githubusercontent.com/8029544/112415617-1c515a80-8cf2-11eb-95e0-817f0e8213ba.png)

### Automatically rotate certificates

- ✅ [A] Remove need to distribute `rootCA.pem`, instead using OS-default trust anchors
    - Automatically generate and rotate `rootCA.pem`
    - Remove support for self-signed certs, instead defaulting to Let's Encrypt
    - Preserve support for specific issuers, e.g. [DigiCert via ACME protocol](https://knowledge.digicert.com/solution/Configure-cert-manager-and-DigiCert-ACME-service-with-Kubernetes.html)
    - [maybe] Preserve support for non-cloud-native certs by keeping AWS-specific functionality but defaulting to not using it
- ✅ [B] Remove need for separate `fluentd.pem` by moving fluentd functionality inside the Orc8r proxy
- ✅ [C] Automatically generate and rotate `controller.crt`

Based on these notes, a reasonable cert rotation schedule could be

- `rootCA.pem` n/a, no longer self-signed
- `fluentd.pem` n/a, no longer necessary
- `controller.crt` every 1 year

### Automatically rotate keys

- ✅ [D] Automatically generate and rotate ephemeral keys, with the following effects
    - `bootstrapper.key`: any gateways mid-bootstrap will fail and automatically restart the process
- ✅ [E] Automatically generate and rotate non-ephemeral keys, with the following effects
    - `controller.key`: no additional effect above rotating corresponding cert
    - `certifier.pem`:
        - All operator certificates will need to be renewed
        - All gateways will need to re-bootstrap

A note on invalidation of operator certs

- The NMS operator cert can be automatically regen'd by cert-manager
- Admin operator certs already default to a [validity period of 1 year](https://github.com/magma/magma/blob/509e5b4bcdb68b0021044fb1dc8cf47554446b8f/orc8r/cloud/go/tools/accessc/handlers/common_add.go#L52-L53), so this rotation won't cause any additional unexpected breakages -- operators will still need to renew their operator certs periodically

Based on these side effects, a reasonable key rotation schedule could be

- `bootstrapper.key` every 1 month (used frequently and directly)
- `controller.key` every 1 year
- `certifier.key` every 3 years

### No-touch secrets management

Based on the recommendations so far, Orc8r will also resolve

- ✅ [F] Automatically generate all necessary secrets at install-time
- ✅ [G] Secrets management handled in K8s, agnostic to cloud provider

Finally, when secrets are automatically updated, the corresponding deployments will need to be rolled. Thankfully, [Helm provides a mechanism for automatically rolling deployments based on config/secret changes](https://helm.sh/docs/howto/charts_tips_and_tricks/#automatically-roll-deployments).

## Solution: backward compatibility

By default, we want to use non-self-signed certs (Let's Encrypt by default, with support for e.g. DigiCert via ACME).

However, for backward compatibility and certain edge use-cases, we want to preserve the option to deploy with 10-year, static, self-signed certs. So these will still be managed by cert-manager (rather than AWS), but we won't rotate etc.

Ideally, it should be *super straightforward* for deployers to understand the difference between these two modes, as well as to choose between them -- e.g. *not* by hand-editing a series of values.yml lines.

## Followup tasks

- [ ] Generate new/alternate secrets via cert-manager https://github.com/magma/magma/pull/5584
- [ ] Ensure gateways can connect without specifying rootCA.pem
    - Ref: https://nghttp2.org/documentation/nghttpx-howto.html#specifying-additional-ca-certificate
    - May need to edit nghttpx conf https://github.com/magma/magma/blob/509e5b4bcdb68b0021044fb1dc8cf47554446b8f/orc8r/gateway/configs/templates/nghttpx.conf.template#L32
    - For now, this step will likely be manual validation (we're still spinning up an e2e control-plane integ test functionality)
- [ ] Move fluentd inside Orc8r proxy
    - Removes need for separate `fluentd.pem`
    - @hcgatewood can help with this one, after v1.5 release crunch
- [ ] Update application to consume secrets from cert-manager rather than AWS-managed secrets
    - Roll deployments on changes to secrets
    - Ensure we still support the 10-year + self-signed pattern (via cert-manager, though)
    - Provide guidance to operators on how to manually kick off rotations / revocations
- [ ] Update deploy code (and documentation) to remove AWS-managed secrets

**Comment by karthiksubraveti** [comment]
@hcgatewood 
Thanks for putting this idea up. It sounds like a great idea. Went through cursively and had following questions. 
- If we remove self signed certs, how do we plan to support development and non prod setups ? 
- Automatically rotating keys sounds like a great idea, Can you elaborate on automatic rolling deployments ? Do you know if we there is visibility into this one, operator may not like surprises for pods to restart during unplanned maintenance window. 
- Also can we swap cert-manager for any other certificate manager?

**Reply by hcgatewood** [reply]
- Support non-prod setups
    - Generally speaking, we can preserve non-default support for manual cert management
    - Staging/test environments: should work same as prod environments -- they need a rootCA.pem either way. The only concern is whether e.g. a test environment is airgapped to a specific LAN. If that's an expected use-case, fall back to manual cert management
    - Dev environments: not a concern for now until we go Minikube-first for local dev, at which point we can determine a solution. Worst-case, fall back to manual
- Rolling deployments
    - Unexpected pod restarts are already a requirement -- e.g. if the cloud provider needs to evict a node, etc. So I think infrequent pod restarts (order of months for every service except bootstrapper -- which is probably same frequency as existing unexpected restarts) would be acceptable. If not, operators can configure their orc8r-certs charts with longer rotation intervals
- Flexibility to pivot from cert-manager
    - While not drop-in replacements, if e.g. cert-manager project died we could feasibly pivot to [kube-cert-manager](https://github.com/kelseyhightower/kube-cert-manager) or [certbot](https://certbot.eff.org/). Neither seems as well-suited to the task as cert-manager, but if necessary they could be swapped in. I don't think it's a big concern at this point, since cert-manager's a CNCF sandbox project at the moment, and seems like it fills a niche

**Comment by hcgatewood** [comment]
Based on some feedback, I think it will be important to preserve existing functionality (long-term self-signed certs) as a non-default fallback. I'll update the proposal in the next day or two to make this approach explicit

**Reply by ShubhamTatvamasi** [reply]
Right now on the PR also the creation of the **certs** helm chart is set to **false**, so it's optional for now.

**Reply by hcgatewood** [reply]
Yeah makes sense. I think we can get away with *always* using the cert-manager approach in your PR (fully replacing existing mechanism), we'll just need to be careful to straightforwardly support the existing pattern (self-signed + 10-year validity period for trust anchors). And then we can default to Let's Encrypt + modern rotation schedule

**Reply by hcgatewood** [reply]
I'll also add implementation steps, where the first step will be getting your existing PR merged 🙂

**Reply by ShubhamTatvamasi** [reply]
Awesome. I know the next task would be using all these secrets on pods. I will take up that once this PR has been merged.

**Reply by hcgatewood** [reply]
Updated the discussion with followup tasks and some additional details. Can you update your open PR https://github.com/magma/magma/pull/5584 to accommodate the changes around e.g. defaulting to Let's Encrypt, rotation timings per secret, supporting easy configurable options between default and self-signed, etc

Once it's updated request me as a reviewer / msg me on Slack and we can get it merged 🙂

**Reply by ShubhamTatvamasi** [reply]
Updated the PR and added https://github.com/magma/magma/pull/5584#issuecomment-812355679.
